### PR TITLE
[feature] Added ability to create new developer sandboxes

### PR DESCRIPTION
**Pull request by 64blit**
commit 36c8d03d1476463a793b5267ddc8e35efa8bd1db
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 19:02:28 2024 -0700

    [wip] Cleaned up logging and removed endpoint disconnection

commit fd0ae7f63a463fbaa3bef5c8b10ac1831f809c4d
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 19:01:47 2024 -0700

    [bugfix] fixed JSON body structure for setting manifest/models

commit 1781f4a5ea73f07a2ba75b384dbaee4da55c0bdc
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 18:24:14 2024 -0700

    [wip] Added missing sandbox check for getting/setting mainfest, models, popComp,transform,

commit 4060c6ca5b9204f4aed7b8f842a7e53cdf978afb
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 18:11:02 2024 -0700

    [wip] Added a sandbox checkbox

commit ae9e8299b2e57b9cdaa9553e2276c34e3b5ddbd8
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 18:10:33 2024 -0700

    [wip] Refactored logic and naming to match changePopComp

commit c723392f0248f11aef49c8708e37c530e336c16a
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 17:37:19 2024 -0700

    [wip] Added more logging to listModels. Issue: empty fetch results

commit 3cab7a9f6115d54f0b4d5f881e9a37a1ffadabaf
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Wed May 8 17:36:05 2024 -0700

    [wip] Added static test files for the sanbox feature

commit d71434f2936151ea4dfeff13ac81895e49374ae8
Author: Edmund Dao <61922101+64blit@users.noreply.github.com>
Date:   Tue May 7 22:34:48 2024 -0700

    [feature] Added sandbox support. Note, a potential _pipeline bug is present here

### CODE REVIEW

**Comment by 64blit** [comment]
I noticed in `endpoint.ts` the pipeline json data needed to be extracted from inside an array before it was accessible as expected. 

I can remove the comment and add some checks if this is from an api update - just let me know 👍 
```
                // ////////////////////////////////////////////////////////
                //
                //
                //
                // NOTE: json()[0] is a hack that should be removed
                //        
                //
                //
                // ////////////////////////////////////////////////////////
                this._pipeline = (await response.json())**[0]** as Pipeline
```

File: src/eyepop/endpoint.ts [hunk 1]
```diff
@@ -624,44 +809,78 @@ export class Endpoint {
...
+                // ////////////////////////////////////////////////////////
+                //
+                //
+                //
+                // NOTE: json()[0] is a hack that should be removed
+                //        
+                //
+                //
+                // ////////////////////////////////////////////////////////
+                this._pipeline = (await response.json())[0] as Pipeline
```
Comment (64blit) [inline src/eyepop/endpoint.ts:876]: Here is the culprit

File: examples/web/static/sandbox.js [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    console.warn = (...args) =>
+    {
+        consoleOutputInput.value += "--Warning--: " + args.join(" ") + "\n";
+    };
+
+    consoleOutputInput.value = "Console Output: \n";
+
+    modelManifestInput.value = JSON.stringify(
+        [
+            { "authority": "legacy", "manifest": "https://s3.amazonaws.com/models.eyepop.ai/releases/legacy/1.2.0/manifest.json" },
```
Comment (tschulz) [inline examples/web/static/sandbox.js:40]: I like the example in general, but it exposes a lot of sausage-making and we don't intend to make this public. Please remove the example.

File: src/eyepop/endpoint.ts [hunk 3]
```diff
@@ -143,6 +144,163 @@ export class Endpoint {
         }
     }
 
+    public async manifest(): Promise<any> {
```
Comment (tschulz) [inline src/eyepop/endpoint.ts:147]: You can define a type for the manifest entries in `types.ts` and preserve the illusion of type safety. I found it helpful when staying in the typescript world. This is a benefit for us internally (not so much for our customers). In go this is:

```
// swagger:model
type SourcesEntry struct {
	Authority string `json:"authority"`
	Manifest  string `json:"manifest"`
}
```

File: src/eyepop/endpoint.ts [hunk 4]
```diff
@@ -143,6 +144,163 @@ export class Endpoint {
...
+        }
+        const data = await response.json();
+        if (!data) {
+            console.warn("getManifest: no data");
+        }
+        return data;
+    }
+
+    
+    public async models(): Promise<string|void> {
```
Comment (tschulz) [inline src/eyepop/endpoint.ts:180]: Same here, a type for a model description please. In Go this is

```
// swagger:model
type ModelInstanceDesc struct {
	Id          string      `json:"id"`
	ModelId     string      `json:"model_id"`
	Dataset     string      `json:"dataset"`
	Version     string      `json:"version"`
	ModelFormat ModelFormat `json:"format"`
	ModelType   ModelType   `json:"type"`
	Size        uint64      `json:"size"`
}
```

File: src/eyepop/endpoint.ts [hunk 5]
```diff
@@ -143,6 +144,163 @@ export class Endpoint {
...
+                headers: headers
+            });
+        });
+        if(!response.ok) {
+            const message = await response.text();
+            return Promise.reject(`Unexpected status ${response.status}: ${message}`);
+        }
+    }
+
+    public async loadModel(model: JSON): Promise<string|void> {
```
Comment (tschulz) [inline src/eyepop/endpoint.ts:243]: ditto, the in parameter should be typed. The go version:
```
// swagger:model
type ModelInstanceDef struct {
	ModelId        string      `json:"model_id"`
	Dataset        string      `json:"dataset"`
	Version        string      `json:"version"`
	ModelFormat    ModelFormat `json:"format"`
	ModelType      ModelType   `json:"type"`
}

// swagger:enum ModelFormat
type ModelFormat string

const (
	TENSORFLOW_LITE      ModelFormat = "TensorFlowLite"
	TENSORFLOW_GRAPH_DEF ModelFormat = "TensorFlowGraphDef"
	TORCHSCRIPT          ModelFormat = "TorchScript"
	TORCHSCRIPT_CPU      ModelFormat = "TorchScriptCpu"
	TORCHSCRIPT_CUDA     ModelFormat = "TorchScriptCuda"
)
// swagger:enum ModelType
type ModelType string

const (
	FLOAT16 ModelType = "float16"
	FLOAT32 ModelType = "float32"
	INT8    ModelType = "int8"
	UINT8   ModelType = "uint8"
)

```

File: src/eyepop/endpoint.ts [hunk 6]
```diff
@@ -143,6 +144,163 @@ export class Endpoint {
...
+        });
+        
+        if(!response.ok) {
+            const message = await response.text();
+            return Promise.reject(`Unexpected status ${response.status}: ${message}`);
+        }
+        return await response.json();
+    }
+
+    public async unloadModel(modelId: string): Promise<string|void> {
```
Comment (tschulz) [inline src/eyepop/endpoint.ts:276]: `Promise<void>` instead of `Promise<string|void>`

File: src/eyepop/endpoint.ts [hunk 7]
```diff
@@ -624,44 +809,78 @@ export class Endpoint {
...
+                };
+                // create a sandbox if needed
+                if (this._options.isSandbox && this._sandboxId === null) {
+                    const createSandboxUrl = `${this._baseUrl}/sandboxes`;
+                    this._requestLogger.debug('before POST %s', createSandboxUrl);
+                    const response = await this._client.fetch(createSandboxUrl, {
+                        method: 'POST',
+                        headers: headers
+                    });
+                    const responseJson = await response.json();
```
Comment (tschulz) [inline src/eyepop/endpoint.ts:829]: check for errors, i.e. `response.status <> 200`

File: src/eyepop/endpoint.ts (hunk 1 above)
Comment (tschulz) [reply src/eyepop/endpoint.ts:876]: not sure why we need this

**Review by tschulz — CHANGES_REQUESTED** [review]
Please add a basic unit test that tests the good path of the new methods. Use this as a template:

https://github.com/eyepop-ai/eyepop-sdk-node/blob/main/tests/eyepop/endpoint.connect.transient.test.ts

**Comment by tschulz** [comment]
@64blit reminder to address my requests and get this merged too

File: src/eyepop/endpoint.ts (hunk 1 above)
Comment (64blit) [reply src/eyepop/endpoint.ts]: I was using the wrong pipeline url - fix found here:
https://github.com/eyepop-ai/eyepop-sdk-node/pull/30/commits/e961cd05c30eeb0c986d987e8d471ff8881a6869

**Comment by 64blit** [comment]
Fixes added and sanbox test passes, it was actually an issue with missing routes.

However, two tests are now failing:

Test Suites: 2 failed, 4 passed, 6 total
Tests:       2 failed, 11 passed, 13 total
Failed tests:
  ● EyePopSdk endpoint module auth and connect for transient popId › EyePopSdk changePostTransform transient
  ● EyePopSdk endpoint module auth and connect › EyePopSdk re-auth on 401

**Pull request closed without merging**